### PR TITLE
ci(workflows): set up code coverage

### LINKED ISSUE #169: Set up test coverage
As a Code PushUp contributor, I want to know how well the codebase is covered with tests. The code coverage will be set up as part of the recent efforts on QA and in the future might be included as a separate plugin of the CLI.

## Implementation details
- Every type of test will have its own type of coverage:
   - E2E tests will not have code coverage as they mostly test critical happy paths.
   - Integration tests will have the least granular coverage (function coverage) as they cover a bigger scope.
   - Unit tests should have at least branch coverage (every possible branching logic is tested).
- Setup [CodeCov ](https://about.codecov.io).

### PR DESCRIPTION
**Pull request by Tlacenka**
Closes #169 

In this PR, I set up the code coverage to run in our pipelines.

- Coverage results can be displayed by adding `--coverage` to any test `nx` command (was already working).
- I included coverage output in logs of every CI run (e.g. [here](https://github.com/code-pushup/cli/actions/runs/6893655166/job/18753721481?pr=278)).
- I set up [Codecov](https://about.codecov.io/) for our organisation (currently waiting for integration approval by an organisation owner).
- I created a separate pipeline taking results from all `nx` libraries to upload them and hopefully create a badge.
- The deprecated vitest coverage package [c8](https://www.npmjs.com/package/@vitest/coverage-c8) package was replaced by v8.

image

### CODE REVIEW

File: packages/utils/project.json [hunk 1]
```diff
@@ -36,13 +36,8 @@
...
       "options": {
-        "passWithNoTests": true
-      },
-      "configurations": {
-        "ci": {
-          "ci": true,
-          "codeCoverage": true
-        }
+        "passWithNoTests": true,
+        "reportsDirectory": "../../coverage/packages/utils"
```
Comment (BioPhoton) [inline packages/utils/project.json:40]:
```suggestion
        "reportsDirectory": "../../dist/coverage/utils"
```

Comment (BioPhoton) [reply packages/utils/project.json:40]: Would it be beter to have all in dist?

**Review by BioPhoton — DISMISSED** [review]
Very nice PR!! Love to see the charts in the future ❤️

File: packages/utils/project.json (hunk 1 above)
Comment (matejchalk) [reply packages/utils/project.json:40]: I would keep it in the default Nx location, it's also gitignored. :man_shrugging:

File: .github/workflows/code-coverage.yml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        env:
+          files: ./coverage/packages/utils/coverage-final.json
+          flags: utils
+          CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
+      - name: Upload testing-utils coverage reports to Codecov
+        uses: codecov/codecov-action@v3
+        env:
+          files: ./coverage/testing-utils/coverage-final.json
+          flags: testing-utils
+          CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
```
Comment (matejchalk) [inline .github/workflows/code-coverage.yml:61]: Good enough for now, but in the future we should make this dynamic. Should be possible with a GitHub Actions matrix :thinking:

**Review by matejchalk — APPROVED** [review]
> currently waiting for integration approval by an organisation owner

Approved :heavy_check_mark: 

Repo is here: https://app.codecov.io/gh/code-pushup/cli

File: packages/utils/project.json (hunk 1 above)
Comment (matejchalk) [reply packages/utils/project.json:40]: I would keep it in the default Nx location, it's also gitignored. 🤷‍♂️

File: .github/workflows/code-coverage.yml (hunk 2 above)
Comment (Tlacenka) [reply .github/workflows/code-coverage.yml]: Hopefully done

File: packages/utils/project.json (hunk 1 above)
Comment (Tlacenka) [reply packages/utils/project.json]: Seems to be ignored so it should be okay.

**Review by matejchalk — APPROVED** [review]
Great work :+1:

**Pull request merged**